test_cluster_detection.py

```diff
@@ -0,0 +1,468 @@
+import pytest
+import sklearn.datasets
+
+from modules.cluster_estimation import cluster_estimation
+from modules import detection_in_world
+
+MIN_TOTAL_POINTS_THRESHOLD = 100
+MIN_NEW_POINTS_TO_RUN = 10
+CENTER_BOX_SIZE = 500
+RANDOM_STATE = 0
+
+
+@pytest.fixture()
+def cluster_model():
+    model_created, model = cluster_estimation.ClusterEstimation.create(
+        MIN_TOTAL_POINTS_THRESHOLD,
+        MIN_NEW_POINTS_TO_RUN,
+        RANDOM_STATE,
+    )
+    assert model_created
+    yield model
+
+
+def generate_cluster_data(n_samples_per_cluster: "list[int]", cluster_standard_deviation: int) \
+    -> "tuple[list[detection_in_world.DetectionInWorld], np.array, list[np.array]]":
+    """
+    Returns a list of points (DetectionInWorld objects) with specified points per cluster
+    and standard deviation.
+
+    PARAMETERS
+    ----------
+    n_samples_per_cluster: list[int]
+        List corresponding to how many points to generate for each generated cluster
+        ex: [10 20 30] will generate 10 points for one cluster, 20 points for the next, 
+        and 30 points for the final cluster.
+
+    cluster_standard_deviation: int
+        The standard deviation of the generated points, bigger
+        standard deviation == more spread out points.
+
+    RETURNS
+    -------
+    detections: list[detection_in_world.DetectionInWorld]
+        List of points (DetectionInWorld objects).
+
+    labels: np.array
+        Labels for each point identifying which cluster center the point it belongs to.
+
+    cluster_positions: list[np.array]
+        Coordinate positions of each cluster centers.
+    -------
+    """
+    # .make_blobs() is a sklearn library function that returns a tuple of two values
+    # first value is ndarray of shape (2, total # of samples) that gives the (x,y) 
+    # coordinate of generated data points
+    # second value is the integer labels for cluster membership of each generated point.
+
+    generated_points, labels, cluster_positions = sklearn.datasets.make_blobs(
+        n_samples=n_samples_per_cluster,
+        n_features=2,
+        cluster_std=cluster_standard_deviation,
+        center_box=(0, CENTER_BOX_SIZE),
+        random_state=0,
+        return_centers=True,
+    )
+
+    detections = []
+    for point in generated_points:
+        # Placeholder variables to create DetectionInWorld objects
+        placeholder_vertices = np.array([[0, 0], [0, 0], [0, 0], [0, 0]])
+        placeholder_label = 1
+        placeholder_confidence = 0.5
+
+        result, detection_to_add = detection_in_world.DetectionInWorld.create(
+            placeholder_vertices,
+            point,
+            placeholder_label,
+            placeholder_confidence,
+        )
+
+        assert result
+        assert (detection_to_add is not None)
+        detections.append(detection_to_add)
+
+    return detections, labels, cluster_positions.tolist()
+
+
+def generate_points_away_from_cluster(num_points_to_generate: int,
```

Comment:
Make private by adding double underscore prefix.

Reply:
adding double underscore prefix causes runtime error from __function() not being recognized